host/traceh.c

```diff
@@ -0,0 +1,241 @@
+    fprintf(stream, "%s", asctime(lt));
+    fprintf(
+        stream,
+        "https://github.com/Microsoft/openenclave branch:%s\n",
+        OE_REPO_BRANCH_NAME);
+    fprintf(stream, "Last commit:%s\n\n", OE_REPO_LAST_COMMIT);
+}
+
+static void _write_message_to_stream(
+    FILE* stream,
+    bool is_enclave,
+    oe_log_args_t* args)
+{
+#if defined(__linux__)
+    struct timeval time_now;
+    gettimeofday(&time_now, NULL);
+    struct tm* t = gmtime(&time_now.tv_sec);
+#else
+    time_t lt = time(NULL);
+    struct tm* t = localtime(&lt);
+#endif
+
+    oe_thread thread_id = oe_thread_self();
+
+    fprintf(
+        stream,
+        LOGGING_FORMAT_STRING,
+        t->tm_hour,
+        t->tm_min,
+        t->tm_sec,
+#if defined(__linux__)
+        time_now.tv_usec,
+#else
+        0,
+#endif
+        thread_id,
+        (is_enclave ? "E" : "H"),
+        _log_level_strings[args->level],
+        args->message);
+}
+
+static void _log_session_header()
+{
+    if (!_log_file_name)
+    {
+        if (_log_level != OE_LOG_LEVEL_NONE)
+            _write_header_info_to_stream(stdout);
+        return;
+    }
+
+    // Take the log file lock.
+    if (oe_mutex_lock(&_log_lock) == OE_OK)
+    {
+        FILE* log_file = NULL;
+        log_file = fopen(_log_file_name, "a");
+        if (log_file == NULL)
+        {
+            fprintf(stderr, "Failed to create logfile %s\n", _log_file_name);
+            oe_mutex_unlock(&_log_lock);
+            return;
+        }
+
+        _write_header_info_to_stream(log_file);
+        fflush(log_file);
+        fclose(log_file);
+        oe_mutex_unlock(&_log_lock);
+    }
+}
+
+oe_result_t oe_log_enclave_init(oe_enclave_t* enclave)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    log_level_t level = _log_level;
+
+    _initialize_log_config();
+
+    // Populate arg fields.
+    oe_log_filter_t* arg = calloc(1, sizeof(oe_log_filter_t));
+    if (arg == NULL)
+    {
+        result = OE_OUT_OF_MEMORY;
+        goto done;
+    }
+    arg->path = enclave->path;
+    arg->path_len = strlen(enclave->path);
+    arg->level = level;
+    // Call enclave
+    result = oe_ecall(enclave, OE_ECALL_LOG_INIT, (uint64_t)arg, NULL);
+    if (result != OE_OK)
+        goto done;
+
+    result = OE_OK;
+done:
+    return result;
+}
+
+void oe_log(log_level_t level, const char* fmt, ...)
+{
+    if (_initialized)
+    {
+        if (level > _log_level)
+            return;
+    }
+
+    if (!fmt)
+        return;
+
+    oe_log_args_t args;
+    args.level = level;
+    va_list ap;
+    va_start(ap, fmt);
+    vsnprintf(args.message, OE_LOG_MESSAGE_LEN_MAX, fmt, ap);
+    va_end(ap);
+    log_message(false, &args);
+}
+
+// This is an expensive operation, it involves acquiring lock
+// and file operation.
+void log_message(bool is_enclave, oe_log_args_t* args)
+{
+    if (!_initialized)
+    {
+        _initialize_log_config();
+        _log_session_header();
```

Comment:
The header will not be logged when the logging level is OE_LOG_LEVEL_NONE. 
You are seeing the header got logged to the screen even thought user did not explicitly set any  logging level, this is because the default logging level is OE_LOG_LEVEL_ERROR.
As you mentioned, it feels weird to see header when run normal app in above case.
I can change the header logging only on when it's logging to a file.